### PR TITLE
fix(tests): resolve model_ops device args against the test device

### LINKED ISSUE #3894: [TEST] model_ops_v2 factory ops with a top-level `device` kwarg run on CPU, not on the device
### What happened?

Ops declared in the `model_ops_tests` configs with `device` as a **top-level kwarg** of `sample_inputs_func` never execute on the Spyre device. They run entirely on CPU, and the harness's own device assertion is disarmed in the same motion, so the test reports success rather than flagging that the device was never touched.

The reported case is `torch.arange` in `tests/configs/model_ops_tests/gpt-oss-20b_spyre.yaml`, declared as:

```yaml
- name: torch.arange
  sample_inputs_func:
    args:
      - value: 11
    kwargs:
      device: cuda:0
```

Under the reported command this executes `torch.arange(11, device='cpu')` on both the reference and the "device" side.

Three separate pieces interact to produce this, and the third is what makes it silent.

**1. The sample is built once, for CPU, and that collapses the device kwarg.** `tests/models/test_model_ops_v2.py:411-415` builds the sample with a hardcoded CPU target:

```python
cpu_sample: SampleInput = ops_item.build_sample_input(
    seed=seed,
    test_device="cpu",
    SampleInput=SampleInput,
)
```

`InputsEdits.resolved_kwargs` then applies its documented device-replacement rule (`tests/oot_framework/oot_test_config_models.py:1096-1098`):

```python
# 2. device replacement
if k == "device" and test_device is not None and "cuda" in v:
    out[k] = test_device
    continue
```

Because `test_device` is `"cpu"` at this call, the YAML's `cuda:0` resolves to `cpu`. The intended device substitution is never requested:

```
resolved with test_device=cpu   : {'device': 'cpu'}
resolved with test_device=spyre : {'device': device(type='spyre')}
```

**2. The device sample is derived from the CPU one and cannot repair it.** `test_sample` is built by mapping `_to_target_device` over the CPU sample's kwarg *values* (`test_model_ops_v2.py:484`):

```python
kwargs={k: _to_target_device(v) for k, v in cpu_sample.kwargs.items()},
```

`_to_target_device` relocates **only tensors**; its final branch is a bare `return x` for everything else. A device string or `torch.device` object is not a tensor, so it falls straight through unchanged:

```
'cpu'                -> 'cpu'
device(type='cpu')   -> device(type='cpu')
```

The comment above that line notes kwargs are "plain values only (dim, dtype, keepdim, etc.)", but `device` is precisely a plain value that *does* need translating — it is the one plain kwarg whose value selects where the op runs.

**3. The guard that should catch a CPU output instead blesses it.** `_is_cpu_output` (`test_model_ops_v2.py:296-303`) keys off the already-corrupted resolved value:

```python
device_str = str(test_sample.kwargs.get("device", ""))
return device_str == "cpu" or (device_str == "" and op_name in _FACTORY_OPS)
```

Since step 1 set `device_str` to `"cpu"`, this returns `True`, `expected_device` becomes `cpu`, and the `_confirm_device` assertion at `:509` passes. Step 1 breaks the input and step 3 lowers the bar to match, so nothing reports a problem.

### What did you expect to happen?

`device: cuda:0` in a `*_spyre.yaml` config should resolve to the Spyre device for the device-side run, so that `torch.ara
nge(11, device=...)` actually allocates and computes on Spyre and is compared against a CPU reference.

Correspondingly, `_is_cpu_output` should only be true when the config **author** wrote `cpu`, so it cannot be satisfied by
 a device kwarg that was mis-resolved upstream.

### How can we reproduce it?

`device: cuda:0` in a `*_spyre.yaml` config should resolve to the Spyre device for the device-side run, so that `torch.ara
nge(11, device=...)` actually allocates and computes on Spyre and is compared against a CPU reference.

Correspondingly, `_is_cpu_output` should only be true when the config **author** wrote `cpu`, so it cannot be satisfied by
 a device kwarg that was mis-resolved upstream.

## How can we reproduce it?

The reported command (also reproduces without `--junit-xml`):

```bash
env bash tests/run_test.sh tests/configs/model_ops_tests/gpt-oss-20b_spyre.yaml -k torch_arange -v
```

Observed result — the op runs on CPU, and the run is reported as passing (`XPASS`, see note under "Anything else"):

```
test_model_ops_v2.py::TestSpyreModelOpsPRIVATEUSE1::test_model_ops_db_torch_arange__45_spyre_float16 XPASS
===================== 234 deselected, 1 xpassed in 26.73s ======================
[torch_oot_device_tests_run] Done. Overall exit code: 0
```

Step 1 in isolation — the device kwarg collapsing to `cpu`:

```bash
python - <<'EOF'
import sys; sys.path.insert(0, 'tests')
import torch
from oot_framework.oot_test_config_models import InputsEdits

ie = InputsEdits(args=[{'value': 11}], kwargs={'device': 'cuda:0'})
print('resolved with test_device=cpu   :', ie.resolved_kwargs(test_device='cpu'))
print('resolved with test_device=spyre :', ie.resolved_kwargs(test_device=torch.device('spyre')))
EOF
```

Output:

```
resolved with test_device=cpu   : {'device': 'cpu'}
resolved with test_device=spyre : {'device': device(type='spyre')}
```

Step 2 in isolation — `_to_target_device`'s non-tensor fallthrough leaves the device value untouched, so the CPU-resolved
kwarg survives into the device sample:

```bash
python - <<'EOF'
import torch

def to_target(x, test_device=torch.device('spyre')):
    if torch.is_tensor(x):
        return x.to(test_device)
    if isinstance(x, (list, tuple)):
        return type(x)(to_target(i) for i in x)
    if isinstance(x, dict):
        return {k: to_target(v) for k, v in x.items()}
    return x   # device string / torch.device hits this

for v in ['cpu', torch.device('cpu')]:
    print(f'{v!r:22} -> {to_target(v)!r}')
EOF
```

Output:

```
'cpu'                  -> 'cpu'
device(type='cpu')     -> device(type='cpu')
```

### Any environmental details we need to know?

- torch: `2.13.0+cpu`
- Python: `3.12.13`
- torch-spyre commit: `be1328a8` (branch `main`)
- pytest: `9.1.1`
- platform: Linux x86_64

### Anything else we need to know?

**Scope — not specific to `arange`.** Any op whose `device` arrives as a top-level kwarg is affected. `device` nested *inside* a tensor spec is handled correctly, because those tensors do go through `_to_target_device`; only the top-level plain kwarg is missed. Counting entries across `tests/configs/model_ops_tests/*_spyre.yaml`:

| config | device-kwarg ops | of which factory ops |
|---|---:|---:|
| `Meta-Llama-3.1-8B-Instruct_spyre.yaml` | 3 | 2 |
| `Ministral-3-14B-Instruct-2512_spyre.yaml` | 3 | 2 |
| `Mistral-Small-3.2-24B-Instruct-2506_spyre.yaml` | 5 | 3 |
| `Qwen2.5-7B-Instruct_spyre.yaml` | 2 | 2 |
| `gemma-4-12B-it_spyre.yaml` | 2 | 2 |
| `gemma-4-26B-A4B-it_spyre.yaml` | 2 | 2 |
| `gpt-oss-20b_spyre.yaml` | 3 | 3 |
| `granite-3.3-8b-instruct_spyre.yaml` | 3 | 2 |
| `granite-4.1-8b_spyre.yaml` | 3 | 2 |
| **total** | **26** | **20** |

The factory ops involved are the `_FACTORY_OPS` set in `test_model_ops_v2.py:199-210` (`torch.arange`, `torch.zeros`, `torch.ones`, `torch.full`, `torch.randn`, `torch.randint`, `torch.empty`, ...).

**Current CI impact is limited, but the coverage is illusory.** These cases are xfail-marked today, so they surface as `XPASS` rather than `PASS` and are not yet asserting a false green. The concern is that they provide no device coverage at all while appearing to, and if the xfail marking is removed a genuine device-side regression in these ops would pass unnoticed.

**Suggested fix direction.** Resolve kwargs per target instead of deriving the device sample from the CPU sample — i.e. build the sample a second time with the real `test_device` so the existing rule 2 substitutes the Spyre device. Then change `_is_cpu_output` to key off the **original YAML** value (did the author write `cpu`?) rather than the resolved one, so it can no longer be satisfied by a mis-resolved device. Fixing only step 1 without step 3 would leave the weakened assertion in place; fixing only step 3 would turn these into hard failures without making them actually run on the device.

### Relevant log output

```Shell
$ env bash tests/run_test.sh tests/configs/model_ops_tests/gpt-oss-20b_spyre.yaml -k torch_arange -v

[torch_oot_device_tests_run] Running serial test
============================= test session starts ==============================
platform linux -- Python 3.12.13, pytest-9.1.1, pluggy-1.6.0
rootdir: /home/yohei/tasks/worktrees/main/torch-spyre
configfile: pyproject.toml
plugins: anyio-4.14.2, xdist-3.8.0
collecting ... collected 235 items / 234 deselected / 1 selected

test_model_ops_v2.py::TestSpyreModelOpsPRIVATEUSE1::test_model_ops_db_torch_arange__45_spyre_float16 XPASS [TAGS = model__gpt-oss-20b_spyre dtype__float16 op__torch_arange platform__x86_64 testtype__trunk constant torch.arange.1_spyre] [100%]
  [INPUT SHAPES]
  arg[0]: value=11

===================== 234 deselected, 1 xpassed in 26.73s ======================

[torch_oot_device_tests_run] Per-file result summary:
  PASS      test_model_ops_v2.py                                  1 xpassed in 26.73s
[torch_oot_device_tests_run] Done. Overall exit code: 0
```

### PR DESCRIPTION
**Pull request by yoheiueda**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

A `device` kwarg — and `torch.to`'s positional destination — selects where an op runs, so it now resolves to the device under test.

`test_model_ops_db` built a single sample with `test_device="cpu"` and derived the device-side sample from it by mapping `_to_target_device` over the values. That helper relocates only tensors, so a plain device value passed straight through and the op ran on CPU. `_is_cpu_output` then read back the same `"cpu"` and expected a CPU result, so `_confirm_device` was satisfied and nothing surfaced.

The fix resolves kwargs a second time against the real `test_device` for the device sample, and adds `InputsEdits.resolved_device_args` for `torch.to`'s positional destination. Tensors stay CPU-built, so both samples hold identical values and the layout-aware `to_spyre()` path in `_to_target_device` is preserved — only the plain device values differ between the two samples.

`_is_cpu_output` needs no change: a config authored `device: cpu` still resolves to `"cpu"` (rule 2 only substitutes on `"cuda"`), so genuinely-CPU ops keep working and the existing assertion becomes meaningful again once it is fed a correctly-resolved device.

v1 (`tests/models/runner.py`) already resolves the device kwarg against the real `test_device`, so this brings v2 in line with it.

#### Which issue(s) this PR is related to:

Fixes #3894

#### Special notes for your reviewer:

**Why `build_cpu_args` is not simply called with the real device.** It passes tensors through `_move_to_test_device`, which would relocate them early and bypass the `arg_spec`-based `to_spyre()` layout path in `_to_target_device`. The substitution therefore has to stay at the plain-value layer, which is why the device kwargs are resolved separately rather than by re-running the whole builder on the device.

**The test result does not change, and that is expected.** `aten.arange` is not implemented on Spyre, so the op now dispatches to the device and falls back to CPU, emitting a new `FallbackWarning: aten.arange.default is falling back to cpu`. That warning is the visible evidence the op reaches the device at all; the fallback returns correct values, so the case stays `XPASS`. This PR restores real coverage rather than flipping a pass/fail.

**Verification.** Full `gpt-oss-20b_spyre.yaml`, before and after:

| | result |
|---|---|
| pre-fix | 46 xfailed, 189 xpassed, exit 0 |
| with fix | 46 xfailed, 189 xpassed, exit 0 |

Resolution behavior, including the negative case that keeps CPU ops working:

```
arange kwargs @spyre : {'device': device(type='spyre')}
authored device: cpu : {'device': 'cpu'}
to positional @spyre : {1: device(type='spyre')}
non-to op            : {}
```

Eager check that the device request now lands:

```
>>> torch.arange(11, device=torch.device('spyre'))
arange on spyre -> spyre:0 torch.int64
values: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

`pre-commit run --files` on both changed files passes (ruff check, ruff format, mypy).

**Follow-up worth considering, not included here.** No current config declares a `device` kwarg with zero positional args, but the kwargs-only branch is wired up the same way so such a config would behave correctly if one is added. Separately, `aten.arange` falling back to CPU on device is the reason this case cannot yet assert a real device computation — enabling it would let the xfail be dropped.

#### Does this PR introduce a user-facing change?

No. Test-framework only; no change to `torch_spyre/` runtime or codegen.

#### Additional note:

Affects 26 op entries across the `*_spyre.yaml` model configs, 20 of them factory ops (`arange`, `zeros`, `ones`, `full`, `randn`, ...). `device` nested inside a tensor spec was already handled correctly — those tensors go through `_to_target_device` — so only the top-level plain value was missed.

### CODE REVIEW

**Comment by yoheiueda** [comment]
@ashokponkumar @anubhavjana @moriohara Cloud you take a look at this PR?

**Review by anubhavjana — APPROVED** [review]

**Review by moriohara — APPROVED** [review]
lgtm
